README.md
# Icelab Assets

An opinionated asset setup for Icelab projects.

## Installation

Add `icelab-assets` as a `devDependency` to your app with one of the below tasks:

```
npm install --save-dev icelab/icelab-assets
yarn add --dev icelab/icelab-assets
```

Then add this config to the `scripts` section of your `package.json` file:

```
"scripts": {
  "start": "icelab-assets start",
  "build": "icelab-assets build",
  "test": "icelab-assets test",
  "create-entry": "icelab-assets create-entry"
}
```

## Usage

Once you’re set up, you’ll have four tasks available that you can run with either `npm run` or `yarn run`:

* `start` will boot a development server at <http://localhost:8080> which will watch your asset files and rebuild on the fly.
* `build` will create a production-ready build for each of your asset entry points.
* `test` will use [jest](https://facebook.github.io/jest/) to run the tests for your assets.
* `create-entry` allows you to create a new asset entry point at a directory path: `yarn run create-entry apps/path/to/the/entry`

By default, `icelab-assets` looks for asset entry points at `apps/**/**/entry.js` and will generate a name for that entry based on its parent directories using this basic formula: `apps/:app_name/**/:entry_name/`. Thus, given the following entry points (and assuming they only return JavaScript):

```
apps/admin/assets/admin/entry.js
apps/admin/assets/inline/entry.js
apps/main/assets/public/entry.js
apps/main/assets/nested/deeply/for/some/reason/critical/entry.js
```

These output files would be generated:

```
admin__admin.js
admin__inline.js
main__public.js
main__critical.js
```

Note: entry names that begin with `inline` are assumed to be files that will eventually be included inline in HTML output. These files thus exclude some dev-related niceties like hot-reloading to avoid clogging up the generated HTML.

## Configuration

### Paths

You can override default path options using the following command line args:

* `--source-path` — defaults to `apps`
* `--build-path` — defaults to `public/assets`
* `--public-path` — defaults to `/assets/`

If you were using this in conjunction with WordPress for example, you might do something like:

```json
"scripts": {
  "start": "icelab-assets start --source-path=wp-content/themes",
  "build": "icelab-assets build --source-path=wp-content/themes",
  "test": "icelab-assets test --source-path=wp-content/themes"
}
```

This would traverse within `wp-content/themes` for any `entry.js` files and use them as the entry points for the build.

### Webpack configurations

You can adjust the webpack configuration for either development or production environments by creating a matching config file in the root of your application:

```
webpack.config.dev.js
webpack.config.prod.js
```

If these files exist, they’ll be merged with the default configuration using the default “smart” strategy from [webpack-merge](https://github.com/survivejs/webpack-merge). You’ll need to match the form of the [existing](config/webpack.config.dev.js) [config](config/webpack.config.prod.js) files to get the to merge matching loaders (including the `include` values for example).

Here’s a custom config that’ll add a `foobar` loader to the pipeline for both CSS and JavaScript:

```
var path = require('path');

module.exports = {
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        // Needs to match the `include` value in the default configuration
        include: path.resolve('apps'),
        use: [{
          loader: 'awesome-loader'
        }]
      },
      {
        test: /\.css$/,
        use: [{
          loader: 'awesome-loader'
        }]
      },
    ],
  }
}
```

This would append the theoretical `awesome-loader` loader to the end of the pipeline in both cases.

### Hashing

Assets are hashed by the `build` script by default. This can be disabled by calling the script with `ASSETS_HASH_FILENAMES=false`:

```
ASSETS_HASH_FILENAMES=false icelab-assets build
```

## CSS

We post-process CSS using [CSS Next](http://cssnext.io) so all the features listed there are available to us. The top-line niceties are:

* Autoprefixing
* CSS variables
* Nested selectors
* Colour functions

We also add support for importing CSS inline using `@import` via [postcss-import](https://github.com/postcss/postcss-import). Something to note about this is that _all_ references in CSS must be made relative to the root of the current entry — this includes `@import` and `url` references:

```css
/**
 * Given a file structure like:
 *
 * /entry-name
 *   /foo
 *      index.js
 *      image.jpg
 *
 * A reference to `image.jpg` must include the parent path.
 */
.foo {
  background-image: url('foo/image.jpg');
}
```

## JavaScript

### Language features and polyfills

We support a superset of the latest JavaScript standard. In addition to [ES6](https://github.com/lukehoban/es6features) syntax features, we also supports:

* [Exponentiation Operator](https://github.com/rwaldron/exponentiation-operator) (ES2016).
* [Async/await](https://github.com/tc39/ecmascript-asyncawait) (ES2017).
* [Object Rest/Spread Properties](https://github.com/sebmarkbage/ecmascript-rest-spread) (stage 3 proposal).
* [Class Fields and Static Properties](https://github.com/tc39/proposal-class-public-fields) (stage 2 proposal).
* [JSX](https://facebook.github.io/react/docs/introducing-jsx.html) and [Flow](https://flowtype.org/) syntax.

Learn more about [different proposal stages](https://babeljs.io/docs/plugins/#presets-stage-x-experimental-presets-).

Note that **the project only includes a few ES6 [polyfills](https://en.wikipedia.org/wiki/Polyfill)**:

* [`Object.assign()`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) via [`object-assign`](https://github.com/sindresorhus/object-assign).
* [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) via [`promise`](https://github.com/then/promise).
* [`fetch()`](https://developer.mozilla.org/en/docs/Web/API/Fetch_API) via [`whatwg-fetch`](https://github.com/github/fetch).

If you use any other ES6+ features that need **runtime support** (such as `Array.from()` or `Symbol`), make sure you are including the appropriate polyfills manually, or that the browsers you are targeting already support them.

#### Top-level entry point import resolution

We’ve added support for top-level resolution of `import`ed modules. This means that you can specify an import as though you are at the top-level of each entry. For example:

```js
// Given a file at ./entry-folder/foo/bar/index.js
export default function bar () {
  // ... do something
}
// From entry-folder/foo/index.js we can do a relative call as usual:
import bar from "./bar";
// ... or scope it from the root of the `entry-folder`
import bar from "foo/bar";
```

### Code splitting

We support code splitting in babel using the [syntax-dynamic-import](http://babeljs.io/docs/plugins/syntax-dynamic-import/) plugin which allows you to, as the name suggests, dynamically import modules. Here’s an example:

```js
// determine-date/index.js
export default function determineDate() {
  import('moment')
    .then(moment => moment().format('LLLL'))
    .then(str => console.log(str))
    .catch(err => console.log('Failed to load moment', err));
}

// other-file/index.js
import determineDate from 'determine-date'
determineDate(); // moment won't be loaded until this line is _executed_
```

You can use this method for imports within your application too, they’re not restricted to external modules.

The build process will automatically split any dynamic imports out into separate chunks, and the bundled JavaScript will load them on the fly without you having to do anything more. You may notice them as `0.abc1234.chunk.js` files in your final build. This should mean smaller initial payloads for apps, but it’s worth considering that separate chunks will potentially take time to load, and you’ll need to have your UI sympathetic to that.

### Code linting and formatting

JS code in your source paths will be linted using [eslint](http://eslint.org). We use the base config from create-react-app, which is released as a separate package called [eslint-config-react-app](https://www.npmjs.com/package/eslint-config-react-app), and then add some small adjustments in each environment:

* In development, we include [prettier](https://github.com/prettier/prettier) to enforce a code style across our projects. This integration will automatically format both JS and CSS whenever you change files.
* In production, we do not include prettier so it will not break your builds. We do however add a `no-debugger` rule to ensure that you can’t push production code that includes debugging lines.

If you’re using an eslint plugin/extension in your editor, you’ll need to configure it to read the `icelab-assets` configuration as its hidden within the package. For Visual Studio code you can add a workspace-specific configuration that looks like this:

```js
// .vscode/settings.json
// Place your settings in this file to overwrite default and user settings.
{
  // Custom eslint config
  "eslint.options": {
    "configFile": "./node_modules/icelab-assets/eslintrc"
  },
  "eslint.nodePath":  "./node_modules/icelab-assets/node_modules"
}
```

Once that’s integrated, you should be able to use eslint’s "Fix all auto-fixable problems" command to fix and format your code with prettier.

## PhantomJS usage

There are some dev-related packages injected into the development build that aren’t relevant in testing environments and these can cause issues with PhantomJS. If you want to exclude them you’ll need to set:

```
ASSETS_ENV=test
```

In your ENV (either using `.env`) or when you start the development server. This is only relevant for *development*, production builds do not include these packages.

## TODOs

- [ ] [Tree shaking doesn’t work at the moment](https://github.com/facebookincubator/create-react-app/pull/1742), alas. Once it’s sorted in `create-react-app` we should be able to pull it in automatically.
- [ ] Enable relative import paths for CSS references (postcss-import only supports root-level resolution).

## Credits

The structure, concept, and most of the code from [create-react-app](https://github.com/facebookincubator/create-react-app) forms the basis for this repo. We still leverage a bunch of stuff from that project so that we’re providing stable and ongoing improvements.



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

# v2.0.8 2019-07-04

* [Enable babel macros support](https://github.com/icelab/icelab-assets/pull/16)

# v2.0.7 2019-06-28

* [Allow consuming apps to override babelrc](https://github.com/icelab/icelab-assets/pull/15)

# v2.0.6 2019-06-24

* Fix missing plugin reference to ResolveEntryModulesPlugin

# v2.0.5 2019-06-24

* Fix deprecation warning in Prettier output

# v2.0.4 2019-06-24

* Bump eslint past affected versions for [CVE](https://snyk.io/vuln/npm:eslint:20180222)

# v2.0.3 2018-06-22

* Fix top-level absolute requires

# v2.0.2 2018-06-05

* Allow `build` to be run without hashing filenames

# v2.0.1 2018-03-20

* Supress warnings in postcss-custom-properties (through postcss-cssnext)

# v2.0.0 2018-03-20

* Breaking change: Add [`postcss-url`](https://github.com/postcss/postcss-url) to the postcss pipeline. Any `url()` references will need to be updated to be relative paths to the file the reference is included in.

# v1.0.6 2018-03-14

* Ensure production CSS is minified.

# v1.0.5 2018-01-10

* Add `.ico` to default list of files included in bundle when *not* explicitly included by other files.

# v1.0.4 2017-12-13

* Bump webpack deps to avoid issue with child cache.

# v1.0.3 2017-10-19

* Set ASSETS_ENV=test to exclude dev-related entry injections. This is useful in fake-browser environments like phantomjs et al.

# v1.0.2 2017-10-18

* Update prettier parser value to avoid deprecation warning

# v1.0.1 2017-09-28

* Update dependencies across the board. Note: this release disables Hot Module Replacement as it’s not supported by the ExtractTextPlugin (and never really was).

# v1.0.0 2017-09-27

* Initial release


config/webpack.config.dev.js
"use strict";

const path = require("path");
const webpack = require("webpack");
const ExtractTextPlugin = require("extract-text-webpack-plugin");
const FriendlyErrorsWebpackPlugin = require("friendly-errors-webpack-plugin");
const WatchMissingNodeModulesPlugin = require("react-dev-utils/WatchMissingNodeModulesPlugin");
const PrettierPlugin = require("prettier-webpack-plugin");
const ResolveEntryModulesPlugin = require("resolve-entry-modules-webpack-plugin");
const atImport = require("postcss-import");
const postcssURL = require("postcss-url");
const cssNext = require("postcss-cssnext");
const getClientEnvironment = require("./env");
const paths = require("./paths");

// Webpack uses `publicPath` to determine where the app is being served from.
// In development, we serve from the full domain we’re booting the app on.
const protocol = process.env.ASSETS_HTTPS === "true" ? "https" : "http";
const host = process.env.ASSETS_HOST || "localhost";
const port = parseInt(process.env.ASSETS_PORT, 10) || 8080;
const serverBase = `${protocol}://${host}:${port}`;
const publicPath = `${serverBase}${paths.publicPath}`;
// Some apps do not use client-side routing with pushState.
// For these, "homepage" can be set to "." to enable relative asset paths.
const shouldUseRelativeAssetPaths = publicPath === "./";
// Get environment variables to inject into our app.
const env = getClientEnvironment();

// ExtractTextPlugin expects the build output to be flat.
// (See https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/27)
// However, our output is structured with css, js and media folders.
// To have this structure working with relative paths, we have to use custom options.
const extractTextPluginOptions = shouldUseRelativeAssetPaths
  ? // Making sure that the publicPath goes back to to build folder.
    { publicPath: Array(cssFilename.split("/").length).join("../") }
  : {};

// This is the development configuration.
// It is focused on developer experience and fast rebuilds.
// The production configuration is different and lives in a separate file.
module.exports = {
  // You may want 'eval' instead if you prefer to see the compiled output in DevTools.
  // See the discussion in https://github.com/facebookincubator/create-react-app/issues/343.
  devtool: "cheap-module-source-map",
  // Specify the context we expect app code to be loaded from
  context: paths.appSrc,
  // These are the "entry points" to our application.
  // This means they will be the "root" imports that are included in JS bundle.
  // We can have multiple entries and so we merge them with some common values to
  // enable "hot" CSS and auto-refreshes for JS.
  entry: paths.appEntries.reduce((output, entry) => {
    const [name, location] = entry;
    // If the entry name begins with `inline` we consider it inline
    const inline = /__inline/.test(name);
    // Test env?
    const testEnv = process.env.ASSETS_ENV && process.env.ASSETS_ENV === "test"
    // Build up the definition for each entry
    let entryDefinition = [];
    // Skip hot-reload and other unnecessary things for "inline" assets
    if (!inline && !testEnv) {
      entryDefinition = entryDefinition.concat([
        // Add our custom dev niceties:
        // - Supressing the output of the webpack-dev-server and hot-module-resolveLoader
        //   in the console. No more [WDS] and [MHR] messages.
        // - Automatically reloading extracted CSS
        require.resolve("./devNiceties"),
        // When you save a file, the client will either apply hot updates (in case
        // of CSS changes), or refresh the page (in case of JS changes). When you
        // make a syntax error, this client will display a syntax error overlay.
        // the line below with these two lines if you prefer the stock client:
        require.resolve("webpack-dev-server/client") + `?${serverBase}`,
        // Disabled until the ExtractText plugin supports HMR
        // https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/592
        // require.resolve("webpack/hot/dev-server")
      ]);
    }
    // Add these for all entries
    entryDefinition = entryDefinition.concat([
      // We ship a few polyfills by default:
      require.resolve("./polyfills"),
      // Finally, this is where the code lives for this entry
      location
      // We include the app code last so that if there is a runtime error during
      // initialization, it doesn't blow up the WebpackDevServer client, and
      // changing JS code would still trigger a refresh.
    ]);
    output[name] = entryDefinition;
    return output;
  }, {}),
  output: {
    // Next line is not used in dev but WebpackDevServer crashes without it:
    path: paths.appBuild,
    // Add /* filename */ comments to generated require()s in the output.
    pathinfo: true,
    // This does not produce real files. It's just the virtual path that is
    // served by WebpackDevServer in development.
    filename: "[name].js",
    chunkFilename: "chunk.[name].js",
    // This is the URL that app is served from. We use "/" in development.
    publicPath: publicPath
  },
  resolve: {
    // This allows you to set a fallback for where Webpack should look for modules.
    // We read `NODE_PATH` environment variable in `paths.js` and pass paths here.
    // We placed these paths second because we want `node_modules` to "win"
    // if there are any conflicts. This matches Node resolution mechanism.
    // https://github.com/facebookincubator/create-react-app/issues/253
    modules: ["node_modules"]
      .concat(paths.nodePaths),
    // These are the reasonable defaults supported by the Node ecosystem.
    // We also include JSX as a common component filename extension to support
    // some tools, although we do not recommend using it, see:
    // https://github.com/facebookincubator/create-react-app/issues/290
    extensions: [".js", ".json", ".jsx"]
  },
  // Resolve loaders (webpack plugins for CSS, images, transpilation) from the
  // directory of `react-scripts` itself rather than the project directory.
  resolveLoader: {
    modules: [paths.ownNodeModules, paths.appNodeModules]
  },
  module: {
    rules: [
      // Disable require.ensure as it's not a standard language feature.
      { parser: { requireEnsure: false } },
      // First, run the linter.
      // It's important to do this before Babel processes the JS.
      {
        test: /\.(js|jsx)$/,
        enforce: "pre",
        use: [
          {
            // Point ESLint to our predefined config.
            options: {
              configFile: path.join(__dirname, "../eslintrc"),
              useEslintrc: false,
              // Automatically fix eslint-issues (i.e., run through prettier)
              fix: true
            },
            loader: "eslint-loader"
          }
        ],
        include: paths.appSrc
      },
      // ** ADDING/UPDATING LOADERS **
      // The "url" loader handles all assets unless explicitly excluded.
      // The `exclude` list *must* be updated with every change to loader extensions.
      // When adding a new loader, you must add its `test`
      // as a new entry in the `exclude` list for "url" loader.

      // "file" loader makes sure those assets get served by WebpackDevServer.
      // When you `import` an asset, you get its (virtual) filename.
      // In production, they would get copied to the `build` folder.
      {
        exclude: [/\.html$/, /\.(js|jsx)$/, /\.css$/, /\.json$/],
        use: [
          {
            loader: "file-loader",
            options: {
              name: "[path][name].[ext]"
            }
          }
        ]
      },
      // Process JS with Babel.
      {
        test: /\.(js|jsx)$/,
        include: paths.appSrc,
        use: [
          {
            loader: "babel-loader",
            options: {
              babelrc: true,
              presets: [require.resolve("babel-preset-react-app")],
              // This is a feature of `babel-loader` for webpack (not Babel itself).
              // It enables caching results in ./node_modules/.cache/babel-loader/
              // directory for faster rebuilds.
              cacheDirectory: true
            }
          }
        ]
      },
      // The notation here is somewhat confusing.
      // "postcss" loader applies autoprefixer to our CSS.
      // "css" loader resolves paths in CSS and adds assets as dependencies.
      // "style" loader normally turns CSS into JS modules injecting <style>,
      // but unlike in development configuration, we do something different.
      // `ExtractTextPlugin` first applies the "postcss" and "css" loaders
      // (second argument), then grabs the result CSS and puts it into a
      // separate file in our build process. This way we actually ship
      // a single CSS file in production instead of JS code injecting <style>
      // tags. If you use code splitting, however, any async bundles will still
      // use the "style" loader inside the async code so CSS from them won't be
      // in the main CSS file.
      {
        test: /\.css$/,
        // Note: this won't work without `new ExtractTextPlugin()` in `plugins`.
        use: ExtractTextPlugin.extract(
          Object.assign(
            {
              fallback: "style-loader",
              use: [
                {
                  loader: "css-loader",
                  options: {
                    importLoaders: 1
                  }
                },
                {
                  loader: "postcss-loader",
                  options: {
                    ident: "postcss", // https://webpack.js.org/guides/migrating/#complex-options
                    plugins: () => [
                      // Add module-like @import support to our CSS. This sets the context for all imports
                      // to be the base entry point.
                      atImport(),
                      // postcss-url "rebases" any `url()` references in CSS to their original relative
                      // position on the filesystem (so that postcss-import doesn't break things)
                      postcssURL(),
                      // cssnext gives us compilation of future-CSS syntax, it also includes autoprefixer
                      // so we don't need to add that separately.
                      cssNext({
                        browsers: [
                          ">1%",
                          "last 4 versions",
                          "Firefox ESR",
                          "not ie < 9" // React doesn't support IE8 anyway
                        ],
                        features: {
                          customProperties: {
                            warnings: false
                          }
                        },
                      })
                    ]
                  }
                }
              ]
            },
            extractTextPluginOptions
          )
        )
      }
      // ** STOP ** Are you adding a new loader?
      // Remember to add the new extension(s) to the "url" loader exclusion list.
    ]
  },
  plugins: [
    // Friendlier error messages
    new FriendlyErrorsWebpackPlugin(),
    // Makes some environment variables available to the JS code, for example:
    // if (process.env.NODE_ENV === 'development') { ... }. See `./env.js`.
    new webpack.DefinePlugin(env.stringified),
    // Expose each entry as a resolution point so that we can resolve from the
    // root of each entry point and avoid relative requires
    new ResolveEntryModulesPlugin(),
    // This is necessary to emit hot updates (currently CSS only):
    // Disabled until the ExtractText plugin supports HMR
    // https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/592
    // new webpack.HotModuleReplacementPlugin(),
    // Format CSS with Prettier. JS is handled through the eslint-loader
    // (so that we can mix in other eslint configuration)
    new PrettierPlugin({
      parser: "css",
      extensions: [".css"]
    }),
    // Note: this won't work without ExtractTextPlugin.extract(..) in `loaders`.
    new ExtractTextPlugin({
      filename: "[name].css"
    }),
    // If you require a missing module and then `npm install` it, you still have
    // to restart the development server for Webpack to discover it. This plugin
    // makes the discovery automatic so you don't have to restart.
    // See https://github.com/facebookincubator/create-react-app/issues/186
    new WatchMissingNodeModulesPlugin(paths.appNodeModules)
  ],
  // Some libraries import Node modules but don't use them in the browser.
  // Tell Webpack to provide empty mocks for them so importing them works.
  node: {
    fs: "empty",
    net: "empty",
    tls: "empty"
  },
  // Turn off performance hints during development because we don't do any
  // splitting or minification in interest of speed. These warnings become
  // cumbersome.
  performance: {
    hints: false
  }
};
